File: alarm-battles/modules/expo-pose-detector/android/src/main/java/expo/modules/posedetector/ExpoPoseDetectorModule.kt
package expo.modules.posedetector

import android.graphics.BitmapFactory
import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.pose.Pose
import com.google.mlkit.vision.pose.PoseDetection
import com.google.mlkit.vision.pose.PoseLandmark
import com.google.mlkit.vision.pose.defaults.PoseDetectorOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

private class ImageDecodeException(uri: String) :
  CodedException("ERR_POSE_IMAGE_DECODE", "Could not decode image at $uri", null)

/**
 * Runs ML Kit Pose Detection against a single still image (from expo-camera's
 * periodic takePictureAsync snapshots), rather than a live camera frame
 * stream. This trades real-time frame-rate detection for a much simpler,
 * more stable integration for a first spike — see src/motion/README.md.
 */
class ExpoPoseDetectorModule : Module() {
  private val detector by lazy {
    val options = PoseDetectorOptions.Builder()
      .setDetectorMode(PoseDetectorOptions.SINGLE_IMAGE_MODE)
      .build()
    PoseDetection.getClient(options)
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoPoseDetector")

    AsyncFunction("detectPose") { imageUri: String, promise: Promise ->
      val path = Uri.parse(imageUri).path ?: imageUri
      val bitmap = BitmapFactory.decodeFile(path)
      if (bitmap == null) {
        promise.reject(ImageDecodeException(imageUri))
        return@AsyncFunction
      }

      val image = InputImage.fromBitmap(bitmap, 0)
      detector.process(image)
        .addOnSuccessListener { pose -> promise.resolve(serializePose(pose)) }
        .addOnFailureListener { e ->
          promise.reject(CodedException("ERR_POSE_DETECTION", e.message, e))
        }
    }

    OnDestroy {
      detector.close()
    }
  }

  private fun serializePose(pose: Pose): Map<String, Any>? {
    if (pose.allPoseLandmarks.isEmpty()) return null

    val landmarks = mutableMapOf<String, Any>()
    for ((type, name) in TRACKED_LANDMARKS) {
      val lm = pose.getPoseLandmark(type) ?: continue
      landmarks[name] = mapOf(
        "x" to lm.position.x.toDouble(),
        "y" to lm.position.y.toDouble(),
        "z" to lm.position3D.z.toDouble(),
        "inFrameLikelihood" to lm.inFrameLikelihood.toDouble()
      )
    }
    return mapOf("landmarks" to landmarks)
  }

  companion object {
    // Keep in sync with src/motion/landmarks.ts POSE_LANDMARK_NAMES.
    private val TRACKED_LANDMARKS = listOf(
      PoseLandmark.NOSE to "nose",
      PoseLandmark.LEFT_SHOULDER to "left_shoulder",
      PoseLandmark.RIGHT_SHOULDER to "right_shoulder",
      PoseLandmark.LEFT_ELBOW to "left_elbow",
      PoseLandmark.RIGHT_ELBOW to "right_elbow",
      PoseLandmark.LEFT_WRIST to "left_wrist",
      PoseLandmark.RIGHT_WRIST to "right_wrist",
      PoseLandmark.LEFT_HIP to "left_hip",
      PoseLandmark.RIGHT_HIP to "right_hip",
      PoseLandmark.LEFT_KNEE to "left_knee",
      PoseLandmark.RIGHT_KNEE to "right_knee",
      PoseLandmark.LEFT_ANKLE to "left_ankle",
      PoseLandmark.RIGHT_ANKLE to "right_ankle"
    )
  }
}

File: alarm-battles/modules/expo-pose-detector/ios/ExpoPoseDetectorModule.swift
import ExpoModulesCore
import MLKitPoseDetection
import MLKitVision
import UIKit

/// Best-effort iOS counterpart to the Android module (see the .kt file for
/// the reasoning behind single-image detection over a live frame stream).
/// This has not been compiled/run — it needs verification against the
/// exact GoogleMLKit/PoseDetection pod version once `pod install` runs.
public final class ExpoPoseDetectorModule: Module {
  private lazy var poseDetector: PoseDetector = {
    let options = PoseDetectorOptions()
    options.detectorMode = .singleImage
    return PoseDetector.poseDetector(options: options)
  }()

  public func definition() -> ModuleDefinition {
    Name("ExpoPoseDetector")

    AsyncFunction("detectPose") { (imageUri: String, promise: Promise) in
      let path = imageUri.hasPrefix("file://") ? String(imageUri.dropFirst("file://".count)) : imageUri
      guard let uiImage = UIImage(contentsOfFile: path) else {
        promise.reject("ERR_POSE_IMAGE_DECODE", "Could not decode image at \(imageUri)")
        return
      }

      let visionImage = VisionImage(image: uiImage)
      visionImage.orientation = uiImage.imageOrientation

      self.poseDetector.process(visionImage) { poses, error in
        if let error = error {
          promise.reject("ERR_POSE_DETECTION", error.localizedDescription)
          return
        }
        guard let pose = poses?.first else {
          promise.resolve(nil)
          return
        }
        promise.resolve(Self.serialize(pose))
      }
    }
  }

  // Keep in sync with src/motion/landmarks.ts POSE_LANDMARK_NAMES.
  private static let trackedLandmarks: [(PoseLandmarkType, String)] = [
    (.nose, "nose"),
    (.leftShoulder, "left_shoulder"),
    (.rightShoulder, "right_shoulder"),
    (.leftElbow, "left_elbow"),
    (.rightElbow, "right_elbow"),
    (.leftWrist, "left_wrist"),
    (.rightWrist, "right_wrist"),
    (.leftHip, "left_hip"),
    (.rightHip, "right_hip"),
    (.leftKnee, "left_knee"),
    (.rightKnee, "right_knee"),
    (.leftAnkle, "left_ankle"),
    (.rightAnkle, "right_ankle"),
  ]

  private static func serialize(_ pose: Pose) -> [String: Any] {
    var landmarks: [String: Any] = [:]
    for (type, name) in trackedLandmarks {
      let lm = pose.landmark(ofType: type)
      landmarks[name] = [
        "x": lm.position.x,
        "y": lm.position.y,
        "z": lm.position.z,
        "inFrameLikelihood": lm.inFrameLikelihood,
      ]
    }
    return ["landmarks": landmarks]
  }
}
